compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/DerivedConfigUtils.cpp

```diff
@@ -32,52 +33,69 @@ getVectorSizeTileSizes(int64_t rank, int64_t innerDimSize, int64_t vectorSize) {
   return tileSizes;
 }
 
+/// Derives the tiles sizes to use based on loop ranges, the number of threads,
+/// and the optimal vector size. If |allowMultiDimCollapse| is true then this
+/// will attempt to split the optimal vector size along multiple complete dims.
+///
+/// For example, with a vector size of 8 and loop ranges of [64 x 2 x 4] this
+/// would return tile sizes of [1, 2, 4] if |allowMultiDimCollapse| is true and
+/// [1, 1, 4] otherwise. If the loop ranges were instead [64 x 4 x 4] this
+/// would give tile sizes of [1, 1, 4] no matter what because we won't be able
+/// to collapse the vector.transfer_read that results from this choice of tile
+/// size.
 static SmallVector<int64_t>
-getThreadTileSizesFromLoopRanges(SmallVector<int64_t> loopRanges,
-                                 int64_t numThreads, int64_t vectorSize) {
+getVectorTileSizesFromLoopRanges(SmallVector<int64_t> loopRanges,
+                                 int64_t numThreads, int64_t vectorSize,
+                                 bool allowMultiDimCollapse = true) {
+  // If any loop ranges are dynamic, default to a simple vector size based
+  // tile size.
   if (llvm::any_of(loopRanges,
                    [](int64_t s) { return ShapedType::isDynamic(s); })) {
```

Comment:
BTW, this could be just
```suggestion
                  &ShapedType::isDynamic)) {
```
I think